File: bundles/gitlab-eclipse-plugin/src/com/gitlab/eclipse/lsp/GitLabLanguageServerProvider.java
package com.gitlab.eclipse.lsp;

import java.net.URI;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.gitlab.eclipse.lsp.GitLabLanguageServerConfigurationParams.CodeCompletion;
import com.gitlab.eclipse.lsp.GitLabLanguageServerConfigurationParams.FeatureFlags;
import com.gitlab.eclipse.lsp.GitLabLanguageServerConfigurationParams.Telemetry;
import com.gitlab.eclipse.preferences.PreferenceConstants;
import com.gitlab.eclipse.preferences.PreferenceInitializer;
import com.gitlab.eclipse.preferences.SecretStringFieldEditor;
import com.gitlab.eclipse.storage.SecretStorage;

import org.eclipse.jface.preference.BooleanFieldEditor;
import org.eclipse.jface.preference.StringFieldEditor;
import org.eclipse.lsp4e.server.ProcessStreamConnectionProvider;
import org.eclipse.lsp4e.server.StreamConnectionProvider;
import org.eclipse.lsp4j.DidChangeConfigurationParams;
import org.eclipse.lsp4j.jsonrpc.messages.Message;
import org.eclipse.lsp4j.jsonrpc.messages.NotificationMessage;
import org.eclipse.lsp4j.jsonrpc.messages.ResponseMessage;
import org.eclipse.lsp4j.services.LanguageServer;


public class GitLabLanguageServerProvider extends ProcessStreamConnectionProvider implements StreamConnectionProvider {
	public static GitLabLanguageServer languageServer;

	// TODO: Extract into a service which registers the language server.
	public static void onDidChangeConfiguration() {
		if (GitLabLanguageServerProvider.languageServer == null) {
			return;
		}
		
		var preferenceStore = PreferenceInitializer.PREFERENCE_STORE;
		var params = GitLabLanguageServerConfigurationParams.builder()
				.baseUrl(preferenceStore.getString(PreferenceConstants.GITLAB_INSTANCE_URL))
				.codeCompletion(new CodeCompletion(true, List.of(), List.of()))
				.featureFlags(FeatureFlags.builder()
						.remoteSecurityScans(false)
						.streamCodeGenerations(preferenceStore.getBoolean(PreferenceConstants.LANGUAGE_SERVER_STREAM_CODE_GENERATIONS))
						.build())
				.ignoreCertificateErrors(preferenceStore.getBoolean(PreferenceConstants.IGNORE_CERTIFICATE_ERRORS))
				.logLevel(preferenceStore.getString(PreferenceConstants.LANGUAGE_SERVER_LOG_LEVEL))
				.telemetry(new Telemetry(
					preferenceStore.getBoolean(PreferenceConstants.TELEMETRY_ENABLED),
					"https://snowplowprd.trx.gitlab.net")
				);

		var secretStorage = new SecretStorage("gitlab.com");
		var personalAccessToken = secretStorage.getSecret("personal_access_token");
		if (personalAccessToken != null) {
			params.token(personalAccessToken);
		}

		GitLabLanguageServerProvider.languageServer.getWorkspaceService().didChangeConfiguration(new DidChangeConfigurationParams(params.build()));
	}

	public GitLabLanguageServerProvider() {
		// TODO: Support configurable language server binary.
		setCommands(List.of(
			"/Users/erran/gitlab-org/editor-extensions/gitlab-lsp/bin/gitlab-lsp-macos-arm64",
			"--stdio"
		));
		// TODO: Make this workspace based.
		setWorkingDirectory("/Users/erran/eclipse-workspace/gitlab-eclipse-plugin/lsp-sandbox");
	}
	
	@Override
	public void handleMessage(Message message, LanguageServer languageServer, URI rootURI) {
		if (GitLabLanguageServerProvider.languageServer == null && languageServer instanceof GitLabLanguageServer lsp) {
			GitLabLanguageServerProvider.languageServer = lsp;
		}

//		System.out.println("message: " + message);
		if (message instanceof NotificationMessage request) {
			switch (request.getMethod()) {
				case "initialized":
					// TODO: Trigger through initialized rpc method?
					onDidChangeConfiguration();
					return;
				default:
					break;
			}
		}

		super.handleMessage(message, languageServer, rootURI);
	}

	@Override
	public Object getInitializationOptions(URI rootUri) {
		return Map.of(
			"extension", Map.of(
				"name", "gitlab-eclipse-plugin",
				"version", "0.1.0-erran"
			),
			"ide", Map.of(
				"name", "gitlab-eclipse-plugin",
				"vendor", "GitLab",
				"version", "0.1.0-erran"
			),
			"folders", List.of(rootUri.toString())
		);
	}
}

File: gitlab-language-server/src/main/kotlin/com/gitlab/eclipse/lsp/GitLabLanguageServerProvider.kt
package com.gitlab.eclipse.lsp

import com.gitlab.eclipse.lsp.GitLabLanguageServerConfigurationParams.CodeCompletion
import com.gitlab.eclipse.lsp.GitLabLanguageServerConfigurationParams.Telemetry
import com.gitlab.eclipse.preferences.PreferenceConstants
import com.gitlab.eclipse.preferences.PreferenceInitializer
import com.gitlab.eclipse.preferences.storage.SecretStorage
import org.eclipse.lsp4e.server.ProcessStreamConnectionProvider
import org.eclipse.lsp4e.server.StreamConnectionProvider
import org.eclipse.lsp4j.DidChangeConfigurationParams
import org.eclipse.lsp4j.jsonrpc.messages.Message
import org.eclipse.lsp4j.jsonrpc.messages.NotificationMessage
import org.eclipse.lsp4j.services.LanguageServer
import java.net.URI

class GitLabLanguageServerProvider : ProcessStreamConnectionProvider(), StreamConnectionProvider {
    init {
        // TODO: Support configurable language server binary.
        commands = listOf(
            "/Users/erran/gitlab-org/editor-extensions/gitlab-lsp/bin/gitlab-lsp-macos-arm64",
            "--stdio"
        )
        // TODO: Make this workspace based.
        workingDirectory = "/Users/erran/eclipse-workspace/gitlab-eclipse-plugin/lsp-sandbox"
    }

    override fun handleMessage(message: Message, languageServer: LanguageServer, rootURI: URI?) {
        if (Companion.languageServer == null && languageServer is GitLabLanguageServer) {
            Companion.languageServer = languageServer
        }

        if (message is NotificationMessage) {
            when (message.method) {
                "initialized" -> {
                    // TODO: Trigger through initialized rpc method?
                    onDidChangeConfiguration()
                    return
                }

                else -> {}
            }
        }

        super<ProcessStreamConnectionProvider>.handleMessage(message, languageServer, rootURI)
    }

    override fun getInitializationOptions(rootUri: URI?): Any? {
        return mapOf(
            "extension" to mapOf(
                "name" to "gitlab-eclipse-plugin",
                "version" to "0.1.0-erran"
            ),
            "ide" to mapOf(
                "name" to "gitlab-eclipse-plugin",
                "vendor" to "GitLab",
                "version" to "0.1.0-erran"
            ),
            "folders" to listOf(rootUri.toString())
        )
    }

    companion object {
        var languageServer: GitLabLanguageServer? = null

        // TODO: Extract into a service which registers the language server.
        fun onDidChangeConfiguration() {
            if (languageServer == null) {
                return
            }

            val params = GitLabLanguageServerConfigurationParams.builder()
                .baseUrl(PreferenceInitializer.PREFERENCE_STORE.getString(PreferenceConstants.GITLAB_INSTANCE_URL))
                .codeCompletion(CodeCompletion(true, listOf(), listOf()))
                .featureFlags(
                    GitLabLanguageServerConfigurationParams.FeatureFlags.builder()
                        .remoteSecurityScans(false)
                        .streamCodeGenerations(PreferenceInitializer.PREFERENCE_STORE.getBoolean(PreferenceConstants.LANGUAGE_SERVER_STREAM_CODE_GENERATIONS))
                        .build()
                )
                .ignoreCertificateErrors(PreferenceInitializer.PREFERENCE_STORE.getBoolean(PreferenceConstants.IGNORE_CERTIFICATE_ERRORS))
                .logLevel(PreferenceInitializer.PREFERENCE_STORE.getString(PreferenceConstants.LANGUAGE_SERVER_LOG_LEVEL))
                .telemetry(
                    Telemetry(
                        PreferenceInitializer.PREFERENCE_STORE.getBoolean(PreferenceConstants.TELEMETRY_ENABLED),
                        "https://snowplow.trx.gitlab.net"
                    )
                )

            val secretStorage = SecretStorage("gitlab.com")
            secretStorage.getSecret("personal_access_token")?.let { personalAccessToken ->
                if (personalAccessToken.isNotBlank())
                params.token(personalAccessToken)
            }

            languageServer!!.workspaceService.didChangeConfiguration(DidChangeConfigurationParams(params.build()))
        }
    }
}